File: src/conversation_indexer/parsers.rs
//! Format-specific export JSON parsers (FR-MP-09 / FR-MP-10 / FR-MP-11).
//!
//! Each parser normalizes its format's shape into `RawMessage` records;
//! classification lives in `types.rs` so mining semantics are shared.

use serde::Deserialize;
use serde_json::Value;

/// Normalized message record shared by all three parsers.
#[derive(Debug, Clone)]
pub struct RawMessage {
    pub participant: String,
    pub timestamp: String,
    pub text: String,
    pub source: String,
}

fn extract_text_parts(value: &Value) -> String {
    // Claude: content is either a string, or an array of blocks with "text".
    // ChatGPT: content is an object { content_type, parts: [String] }.
    match value {
        Value::String(s) => s.clone(),
        Value::Array(blocks) => {
            let mut out = Vec::new();
            for block in blocks {
                if let Some(text) = block.get("text").and_then(|t| t.as_str()) {
                    out.push(text.to_string());
                }
            }
            out.join("\n")
        }
        Value::Object(map) => {
            if let Some(parts) = map.get("parts").and_then(|p| p.as_array()) {
                let mut out = Vec::new();
                for part in parts {
                    if let Some(s) = part.as_str() {
                        out.push(s.to_string());
                    }
                }
                return out.join("\n");
            }
            map.get("text")
                .and_then(|t| t.as_str())
                .unwrap_or("")
                .to_string()
        }
        _ => String::new(),
    }
}

// ---------------------------------------------------------------- Claude

#[derive(Deserialize)]
struct ClaudeExport {
    conversations: Vec<ClaudeConversation>,
}

#[derive(Deserialize)]
struct ClaudeConversation {
    messages: Vec<ClaudeMessage>,
}

#[derive(Deserialize)]
struct ClaudeMessage {
    role: Option<String>,
    text: Option<String>,
    content: Option<Value>,
}

/// Claude export: `{ "conversations": [ { "messages": [ { role, content } ] } ] }`.
pub fn parse_claude_export(content: &str) -> Result<Vec<RawMessage>, String> {
    let export: ClaudeExport =
        serde_json::from_str(content).map_err(|e| format!("not a Claude export: {e}"))?;
    let mut out = Vec::new();
    for conv in export.conversations {
        for msg in conv.messages {
            let text = match (&msg.text, &msg.content) {
                (Some(t), _) => t.clone(),
                (None, Some(c)) => extract_text_parts(c),
                (None, None) => String::new(),
            };
            out.push(RawMessage {
                participant: msg.role.unwrap_or_else(|| "user".to_string()),
                timestamp: String::new(),
                text,
                source: "claude".to_string(),
            });
        }
    }
    Ok(out)
}

// ---------------------------------------------------------------- ChatGPT

#[derive(Deserialize)]
struct ChatGptExport {
    mapping: std::collections::BTreeMap<String, ChatGptNode>,
}

#[derive(Deserialize)]
struct ChatGptNode {
    #[serde(default)]
    message: Option<ChatGptMessage>,
    #[serde(default)]
    mapping: Option<std::collections::BTreeMap<String, ChatGptNode>>,
}

#[derive(Deserialize)]
struct ChatGptMessage {
    author: Option<ChatGptAuthor>,
    #[serde(default)]
    content: Option<Value>,
}

#[derive(Deserialize)]
struct ChatGptAuthor {
    #[serde(default)]
    role: Option<String>,
}

/// ChatGPT export: `{ "mapping": { "<id>": { "message": { author.role,
/// content.parts[] } } } }`. Nested `mapping` subtrees (conversation nodes
/// that wrap child nodes) are walked recursively.
pub fn parse_chatgpt_export(content: &str) -> Result<Vec<RawMessage>, String> {
    let export: ChatGptExport =
        serde_json::from_str(content).map_err(|e| format!("not a ChatGPT export: {e}"))?;
    let mut out = Vec::new();
    collect_chatgpt_nodes(&export.mapping, &mut out);
    Ok(out)
}

fn collect_chatgpt_nodes(
    nodes: &std::collections::BTreeMap<String, ChatGptNode>,
    out: &mut Vec<RawMessage>,
) {
    for node in nodes.values() {
        if let Some(msg) = &node.message {
            if let Some(content) = msg.content.as_ref() {
                let text = extract_text_parts(content);
                if !text.trim().is_empty() {
                    out.push(RawMessage {
                        participant: msg
                            .author
                            .as_ref()
                            .and_then(|a| a.role.clone())
                            .unwrap_or_else(|| "user".to_string()),
                        timestamp: String::new(),
                        text,
                        source: "chatgpt".to_string(),
                    });
                }
            }
        }
        if let Some(nested) = &node.mapping {
            collect_chatgpt_nodes(nested, out);
        }
    }
}

// ---------------------------------------------------------------- Slack

#[derive(Deserialize)]
struct SlackExport {
    #[serde(default)]
    #[allow(dead_code)] // shape marker: distinguishes Slack channel exports
    channel: Option<Value>,
    messages: Vec<SlackMessage>,
}

#[derive(Deserialize)]
struct SlackMessage {
    #[serde(default)]
    r#type: Option<String>,
    #[serde(default)]
    user: Option<String>,
    #[serde(default)]
    text: Option<String>,
}

/// Slack channel export: `{ "channel": {...}, "messages": [ { user, ts,
/// text } ] }`. Works for both full channel files and files.json-style
/// single-message objects that wrap a `message` field.
pub fn parse_slack_export(content: &str) -> Result<Vec<RawMessage>, String> {
    let export: SlackExport =
        serde_json::from_str(content).map_err(|e| format!("not a Slack export: {e}"))?;
    let mut out = Vec::new();
    for msg in export.messages {
        if msg.r#type.as_deref() == Some("message") || msg.text.is_some() {
            out.push(RawMessage {
                participant: msg.user.unwrap_or_else(|| "unknown".to_string()),
                timestamp: String::new(),
                text: msg.text.unwrap_or_default(),
                source: "slack".to_string(),
            });
        }
    }
    Ok(out)
}

File: go/internal/convo/parsers.go
// Format-specific export JSON parsers (FR-MP-09 / FR-MP-10 / FR-MP-11),
// ported from the Rust `conversation_indexer::parsers` module.
//
// Each parser normalizes its format's shape into RawMessage records;
// classification lives in types.go so mining semantics are shared.
package convo

import (
	"bytes"
	"encoding/json"
	"fmt"
	"sort"
	"strings"
)

// RawMessage is the normalized message record shared by all three parsers.
type RawMessage struct {
	Participant string
	Timestamp   string
	Text        string
	Source      string
}

// extractTextParts flattens the three content shapes the exports use:
//
//	Claude:  a string, or an array of blocks with a "text" field
//	ChatGPT: an object { content_type, parts: [string] }
//	other:   an object with a "text" field
//
// It returns an empty string for any other shape.
func extractTextParts(value json.RawMessage) string {
	if len(bytes.TrimSpace(value)) == 0 {
		return ""
	}
	var v any
	if err := json.Unmarshal(value, &v); err != nil {
		return ""
	}
	switch t := v.(type) {
	case string:
		return t
	case []any:
		parts := make([]string, 0, len(t))
		for _, block := range t {
			if m, ok := block.(map[string]any); ok {
				if s, ok := m["text"].(string); ok {
					parts = append(parts, s)
				}
			}
		}
		return joinLines(parts)
	case map[string]any:
		if raw, ok := t["parts"].([]any); ok {
			parts := make([]string, 0, len(raw))
			for _, p := range raw {
				if s, ok := p.(string); ok {
					parts = append(parts, s)
				}
			}
			return joinLines(parts)
		}
		if s, ok := t["text"].(string); ok {
			return s
		}
		return ""
	default:
		return ""
	}
}

func joinLines(parts []string) string {
	out := ""
	for i, p := range parts {
		if i > 0 {
			out += "\n"
		}
		out += p
	}
	return out
}

// ---------------------------------------------------------------- Claude

type claudeExport struct {
	// Conversations is a pointer so a missing root key is distinguishable
	// from an empty array (Rust's serde rejects the former).
	Conversations *[]struct {
		Messages []struct {
			Role    *string         `json:"role"`
			Text    *string         `json:"text"`
			Content json.RawMessage `json:"content"`
		} `json:"messages"`
	} `json:"conversations"`
}

// parseClaudeExport parses `{ "conversations": [ { "messages": [ { role,
// content } ] } ] }`. A message's explicit `text` field wins over its
// `content`; an absent role falls back to "user".
func parseClaudeExport(content []byte) ([]RawMessage, error) {
	var export claudeExport
	if err := json.Unmarshal(content, &export); err != nil {
		return nil, fmt.Errorf("not a Claude export: %w", err)
	}
	if export.Conversations == nil {
		return nil, fmt.Errorf("not a Claude export: missing conversations array")
	}
	var out []RawMessage
	for _, conv := range *export.Conversations {
		for _, msg := range conv.Messages {
			text := ""
			switch {
			case msg.Text != nil:
				text = *msg.Text
			case len(msg.Content) > 0:
				text = extractTextParts(msg.Content)
			}
			participant := "user"
			if msg.Role != nil {
				participant = *msg.Role
			}
			out = append(out, RawMessage{
				Participant: participant,
				Text:        text,
				Source:      "claude",
			})
		}
	}
	return out, nil
}

// ---------------------------------------------------------------- ChatGPT

type chatGPTExport struct {
	Mapping *map[string]chatGPTNode `json:"mapping"`
}

type chatGPTNode struct {
	Message *struct {
		Author *struct {
			Role *string `json:"role"`
		} `json:"author"`
		Content json.RawMessage `json:"content"`
	} `json:"message"`
	Mapping map[string]chatGPTNode `json:"mapping"`
}

// parseChatGPTExport parses `{ "mapping": { "<id>": { "message": {
// author.role, content.parts[] } } } }`. Nested `mapping` subtrees
// (conversation nodes that wrap child nodes) are walked recursively.
func parseChatGPTExport(content []byte) ([]RawMessage, error) {
	var export chatGPTExport
	if err := json.Unmarshal(content, &export); err != nil {
		return nil, fmt.Errorf("not a ChatGPT export: %w", err)
	}
	if export.Mapping == nil {
		return nil, fmt.Errorf("not a ChatGPT export: missing mapping object")
	}
	var out []RawMessage
	collectChatGPTNodes(*export.Mapping, &out)
	return out, nil
}

// sortedKeys returns the node ids in sorted order (Rust used a BTreeMap).
func sortedKeys(m map[string]chatGPTNode) []string {
	keys := make([]string, 0, len(m))
	for k := range m {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	return keys
}

// collectChatGPTNodes walks the mapping in sorted key order and appends one
// record per non-empty message.
func collectChatGPTNodes(nodes map[string]chatGPTNode, out *[]RawMessage) {
	for _, id := range sortedKeys(nodes) {
		node := nodes[id]
		if node.Message != nil && len(node.Message.Content) > 0 {
			if text := extractTextParts(node.Message.Content); strings.TrimSpace(text) != "" {
				participant := "user"
				if node.Message.Author != nil && node.Message.Author.Role != nil {
					participant = *node.Message.Author.Role
				}
				*out = append(*out, RawMessage{
					Participant: participant,
					Text:        text,
					Source:      "chatgpt",
				})
			}
		}
		if len(node.Mapping) > 0 {
			collectChatGPTNodes(node.Mapping, out)
		}
	}
}

// ---------------------------------------------------------------- Slack

type slackExport struct {
	Messages *[]struct {
		Type *string `json:"type"`
		User *string `json:"user"`
		Text *string `json:"text"`
	} `json:"messages"`
}

// parseSlackExport parses `{ "channel": {...}, "messages": [ { user, ts,
// text } ] }` — both full channel files and files.json-style single-message
// objects that wrap a `message` field. Messages are kept when typed
// "message" or when they carry text; the participant falls back to unknown.
//
// NOTE (parity): the Rust SlackMessage struct never deserializes `ts`, so
// every mined Slack timestamp is empty; this port keeps that shape. Extracted
// records are unaffected (classification and topics come from the text).
func parseSlackExport(content []byte) ([]RawMessage, error) {
	var export slackExport
	if err := json.Unmarshal(content, &export); err != nil {
		return nil, fmt.Errorf("not a Slack export: %w", err)
	}
	if export.Messages == nil {
		return nil, fmt.Errorf("not a Slack export: missing messages array")
	}
	var out []RawMessage
	for _, msg := range *export.Messages {
		if (msg.Type != nil && *msg.Type == "message") || msg.Text != nil {
			participant := "unknown"
			if msg.User != nil {
				participant = *msg.User
			}
			text := ""
			if msg.Text != nil {
				text = *msg.Text
			}
			out = append(out, RawMessage{
				Participant: participant,
				Text:        text,
				Source:      "slack",
			})
		}
	}
	return out, nil
}
